aws/efs/cmd/efs-provisioner/filesystemreclaimer.go

```diff
@@ -0,0 +1,151 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"strconv"
+	"syscall"
+
+	"github.com/golang/glog"
+	"github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/allocator"
+	"github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/controller"
+	"github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/gidreclaimer"
+	"k8s.io/kubernetes/pkg/apis/core/v1/helper"
+)
+
+// compile time check to make sure fileSystemReclaimer implements the GIDReclaimer interface
+var _ gidreclaimer.GIDReclaimer = &fileSystemReclaimer{}
+
+func newFileSystemReclaimer(basePath string) *fileSystemReclaimer {
+	return &fileSystemReclaimer{basePath}
+}
+
+type fileSystemReclaimer struct {
+	BasePath string
+}
+
+// Reclaim looks at every top level directory in the basepath and adds its gid to the given gidTable
+func (f *fileSystemReclaimer) Reclaim(classname string, gidtable *allocator.MinMaxAllocator) error {
+	glog.Infof("adding gids for any existing directories under %s to the gid table", f.BasePath)
+
+	entries, err := ioutil.ReadDir(f.BasePath)
+	if err != nil {
+		glog.Errorf("failed to list contents of %v: %v", f.BasePath, err)
+		return err
+	}
+
+	for _, entry := range entries {
+		if !entry.IsDir() {
+			continue
+		}
+
+		mddir := path.Join(f.BasePath, entry.Name())
+
+		md, err := readVolumeMetadata(mddir)
+		if err != nil {
+			glog.Warningf("failed to read volume metadata for %v: %v", mddir, err)
+			continue
+		}
+
+		// if no metadata then it must have been created by another storage class that doesn't have reuseVolumes set since those don't write metadata
+		if md == nil {
+			continue
+		}
+
+		// no GID was previously allocated
+		if md.GID == "" {
+			continue
+		}
+
+		// skip volumes for other storage classes
+		if md.StorageClassName != classname {
+			continue
+		}
+
+		gid, err := strconv.Atoi(md.GID)
+		if err != nil {
+			glog.Errorf("invalid GID value '%v' in metadata for %v", md.GID, mddir)
+			continue
+		}
+
+		_, err = gidtable.Allocate(gid)
+		if err == allocator.ErrConflict {
+			glog.Infof("GID %v found in %v was already allocated for storageclass %v", gid, mddir, classname)
+			continue
+		} else if err != nil {
+			glog.Errorf("failed to store GID %v found in metadata for %v: %v", gid, mddir, err)
+			continue
+		}
+	}
+
+	return nil
```

Comment:
None of the errors above are _actually_ errors as far as the whole process is concerned?

Reply:
Are you asking if the error cases should be stopping the loop and returning an error?  I think not.  This is checking every subdirectory in the basepath for gids that aren't allocated and marking them allocated.  Here was my thinking.  If you see it differently let me know:

```
glog.Errorf("invalid GID value '%v' in metadata for %v", md.GID, mddir)
```

If one directory has a crap GID value in its metadata, the only reason I could conceive of is that someone has been messing around. Report it, but don't make the whole provisioner die and don't fail to allocate as many gids as possible.

```
glog.Errorf("failed to store GID %v found in metadata for %v: %v", gid, mddir, err)
```

Having looked at the code for the Allocate method, the reasons you'd hit this case are `ErrOutOfRange` or `ErrConflict`.
- The `ErrConflict` case is already checked (and if it occurs, then hooray, the gid was already allocated, which is exactly what we're trying to do anyway.  This case will happen when a PVC is deleted and recreated within the same cluster).
- That only leaves `ErrOutOfRange`.  The only reason I could conceive of for `ErrOutOfRange` happening is either someone hacked the metadata (sorry about your luck, don't hack crap if you don't know what you're doing), or the storageclass has been deleted and recreated with a new range.   Kubnernetes won't let you modify a storage class.  It would have to be deleted and recreated with different values (or someone tore down their cluster and put the storageclassback with differnt values, but same difference).  If someone does this, it seems like they also ought to know that it will have consequences.  The actual gid is on the directory and all of the files in the file system, so you already can't just change the range without having to do something to the files to get the permissions on them to be right if you want the gid on the directory to actually be inside of your new range.  If someone actually changed the range but left the file system and metadata alone, then when the PVC is processed, the GID won't have been marked in use in the GID table by reclaim process, but the directory will still be hooked back up using the original GID it had in the first place. However, it won't conflict with another GID for any new PVCs of in the storageclass, because it's outside the range if GIDs now being assigned to new PVCs in that class.  If you think it's worth it I can actually test to make sure that's the case, but I'm pretty sure it is.